internal/linker/linker.go

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2022, The Garble Authors.
+// See LICENSE for licensing information.
+
+package linker
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"embed"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/bluekeyes/go-gitdiff/gitdiff"
+	"github.com/gofrs/flock"
+	"io"
+	"io/fs"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+const (
+	MagicValueEnv = "GARBLE_LNK_MAGIC"
+
+	cacheDirName   = ".garble"
+	versionExt     = ".version"
+	garbleCacheDir = "GARBLE_CACHE_DIR"
+)
+
+var (
+	//go:embed patches/*.patch
+	linkerPatchesFs embed.FS
+
+	linkerPatchesVer string
+	linkerPatches    map[string]string
+
+	baseSrcSubdir = filepath.Join("src", "cmd")
```

Comment:
nit: I would do the patching relative to `src`, not `src/cmd`, because we may well need to apply static patches to bits of std in the future

Reply:
as a bonus, then the patch filenames use package paths, e.g. `cmd/link/internal/ld/pcln.go` is the file `pcln.go` in package `cmd/link/internal/ld`.